### ISSUE TITLE
World Conquest - Map Generation Scoping WML.Variables bug

### ISSUE THREAD

**Issue by AskAardvark** [wesnoth/wesnoth#9463]
### Game and System Information

- **Version:** 
- **OS:** Windows

PR Example: https://github.com/wesnoth/wesnoth/pull/9430/files

### Description of the bug

A slider creating an integer is not pulled into Wesnoth map generation. There are other variables in custom_bonus.lua are also pulling from wml.variables which may or may not be working.

The [Optionswml Wiki](https://wiki.wesnoth.org/OptionWML) states "variables in scenariolevel [lua] tags and prestart events." should load properly.

The PR demonstrates a slider value working correctly in game loading.

### Steps to reproduce the behavior

1. Update client to the commit of the PR: https://github.com/wesnoth/wesnoth/pull/9430/files
2. Attempt to load a singleplayer world conquest map. The # set for custom bonuses is indifferent because of the null reference error.

### Expected behavior

WML.Variables["wml.variables["custom_map_bonus_per_player"] loads from an option slider when bonus_points.lua is run.

### Additional context

Related commit shows this was a known issue but instead a workaround was created: https://github.com/wesnoth/wesnoth/commit/bca7a60852cbbf309ac9b5bf9cc48ddad7910436

[This PR](https://github.com/wesnoth/wesnoth/pull/9430) would like to add a small customization feature to World Conquest requiring addressing the underlying scoping issue.

**Comment by gfgtdf** [comment]
i wouldn't call is a bug because map generation happens clearly before scenariolevel [lua] tags and prestart events, so the quoted line in the wiki page doesn't guarantee that it will work there.

(which doesn't mean that we cannot implement this)

**Comment by CelticMinstrel** [comment]
I would say that variables defined in the scenario (using `[variables]`) should probably be visible to the map generation. Furthermore, OptionWML variables would ideally be added to the scenario in that way. This is just an ideal though… it sounds difficult to implement, since the order of operations is all wrong to support that… given that map generation runs before you even _see_ the OptionWML.

I would also say that it is _expected_ for map generation to run before scenario level `[lua]` tags, as the map generation can potentially generate elements of the scenario itself, which could theoretically include some scenario level `[lua]` tags. So if there is any change made it should only be with regards to `[variables]`.

**Comment by gfgtdf** [comment]
> given that map generation runs before you even see the OptionWML.

this is true for map_generation in [multiplayere] but not in [campaign] iirc, (so if we impleemnt this, this would also only for for [campaign]), it _might_ be as easy as moving this https://github.com/wesnoth/wesnoth/blob/b03caa9205dcea1e37b61d699d86ca5a0f54f64b/src/game_initialization/mp_game_utils.cpp#L62 2 lines up

**Comment by sevu** [comment]
This original intention was to find some way to enable more bonus_points in WC2.

**Comment by sevu** [comment]
> But that said, there is already a separate way of specifying configuration options to the map generator, so for right now you could just use that instead of OptionsWML for these cases.

Can one add from UMC a custom entry in the UI to what is shown behind the map generator settings button?

**Comment by soliton-** [comment]
IMO in general it would be good if map generation can use custom options. It's true that it's a bit complicated since the options and the generated map are shown in the same screen.

One possibility may be to allow marking options as affecting map generation and then if such an option is changed the map is regenerated. Don't know if it'd be useful to also show such options in the map generator settings dialog instead. I fear they'd be easy to miss there.

---

For WC specifically is it important that the bonus generation happens at map generation time? I think it might be easiest to workaround this issue in WC by moving the bonus item handling to prestart or so.

**Comment by gfgtdf** [comment]
thinking about it, we could also move half of it to lua (the first filter), and make the actual placement in the scenaio code

**Comment by gfgtdf** [comment]
> Can one add from UMC a custom entry in the UI to what is shown behind the map generator settings button?

Partly, the (lua) map gnerator code can add whatever its wants to it but its not possible for for example [modification]s to add stuff to it. Also, this button doesnt even show for world Conquest since its a [campaign] and that button only shows for [multiplayer]

**Comment by AskAardvark** [comment]
Bump! I'd love to see this bug prioritized and fixed!
